[Demo] Local-only supabase todo app

Chriztiaan opened the pull request:
This adds a demo project that aims to demonstrate how an app could be built starting from a local-only state and later being "upgraded" to a version that has sync capabilities. It started off as a copy of the existing todo app that was stripped of some of the features. The biggest changes are in `powersync.dart`, login/signup views, and `models/schema.dart` (which illustrates how to update the schema to supporting syncing).

The readme has a basic description of how everything works, with a small mermaid diagram to show how the app switches from the local-only mode to sync mode.

stevensJourney commented on demos/local-only-todolist/README.md:
```diff
@@ -0,0 +1,70 @@
+
+1. `cd demos/local-only-todolist`
+2. `melos prepare`
+3. `cp lib/app_config_template.dart lib/app_config.dart`
```
Is it possible to not require this in order to get started initially? It feels like this is only for the online mode - it would be nice to be able to get the demo started without having to setup the cloud services. It's not super important though.

kobiebotha replied:
Setting up the cloud services is significant friction if they aren't needed

stevensJourney commented on demos/local-only-todolist/lib/models/sync_mode.dart:
```diff
@@ -0,0 +1,48 @@
+  await migrations.migrate(sqliteDb);
+}
+
+Future<bool> getSyncMode() async {
```
Maybe this could be called `getSyncEnabled` (with `setSyncEnabled`) since it's returning (or setting) a boolean state?

stevensJourney commented on demos/local-only-todolist/lib/models/schema.dart:
```diff
@@ -0,0 +1,87 @@
+/// the main 'lists' / 'todos' view name.
+///
+/// For local only, the views become:
+///   online_todos
```
I think this is switched with the online version?

stevensJourney replied:
Never mind, had to do some mental processing for this.

kobiebotha replied:
This caught me too, we should perhaps expand the comment for a bit of hand holding?

Chriztiaan commented:
@kobiebotha @stevensJourney  I've updated the readme, schema comment about switching, and allowed the app to work without specifying credentials in which case login/sign up is disabled.

guillempuche commented:
Good work! Much needed for a seamless experience.
I'll try to implement for my app.

stevensJourney reviewed (dismissed):
Nice!

kobiebotha commented on demos/local-only-todolist/README.md:
```diff
@@ -0,0 +1,74 @@
+# PowerSync + Supabase Flutter Local Only to Online Mode Demo: Todo List App
+
+This demo app is an extension of the Flutter Todo List App that demonstrates how to use the PowerSync SDK for Flutter in an local only way. It allows the user to use the app offline without having to sign in. After signing in data that was kept locally is synced up. For a step-by-step guide, see [here](https://docs.powersync.com/integration-guides/supabase).
```
- of the *Supabase Todo List App
- in *a local only way

We could also make the intro sentence more explicit to help people grok the demo:

`This demo app is an extension of the Flutter Todo List App that demonstrates how to use the PowerSync SDK for Flutter in an local only way without sync capabilities.`

kobiebotha commented on demos/local-only-todolist/README.md:
```diff
@@ -0,0 +1,74 @@
+
+# Explanation
+
+The demo implements local-only and synced modes by using two sets of schema definitions, which can be viewed [here](./lib/models/schema.dart). The app initially starts in local only mode with the offline schema. When the user signs in, the database schema is updated to the online schema, and the data is migrated to enable synchronization.
```
s/local only/local-only (best to be consistent)

Chriztiaan replied:
I only learned about `s/old/new` after reading this message 5 times 😆

kobiebotha commented on demos/local-only-todolist/lib/models/schema.dart:
```diff
@@ -0,0 +1,98 @@
+import 'package:powersync/powersync.dart';
+import 'package:powersync_flutter_local_only_demo/models/sync_mode.dart';
+
+/// This schema design supports an online/local-only workflow by managing data
```
should this be `supports a local-only to online workflow` ?

kobiebotha commented on demos/local-only-todolist/lib/models/schema.dart:
```diff
@@ -0,0 +1,98 @@
+///
+/// - 'todos' and 'lists' refer to the local-only data.
+/// - 'online_todos' and 'online_lists' refer to the data that will be synced
+///    once online, making it clear that these are not currently synchronized.
```
Thinking about it now I actually don't think the name `online_todos` makes it clear that these are not synced, they sound synced. What about `queued_todos`, `pending_upload_todos` or something like that?

rkistner replied:
Just to confirm what we discussed - I like the new prefixes, and feel it makes it a little more more clear what's happening.

Chriztiaan commented:
> Good work! Much needed for a seamless experience. I'll try to implement for my app (Typescript).

Hey @guillempuche, a JS/TS version is on the one in the coming weeks too.

benitav commented on demos/local-only-todolist/README.md:
```diff
@@ -0,0 +1,74 @@
+# PowerSync + Supabase Flutter Local-Only to Online Mode Demo: Todo List App
+
+This demo app is an extension of the Supabase Todo List App that demonstrates how to use the PowerSync SDK for Flutter in a local-only way without sync capabilities. It allows the user to use the app offline without having to sign in. After signing in data that was kept locally is synced up. For a step-by-step guide, see [here](https://docs.powersync.com/integration-guides/supabase).
```
```suggestion
This demo app is an extension of the [Supabase Todo List App](./supabase-todolist/README.md) that demonstrates how to use the PowerSync SDK for Flutter in a local-only way without sync capabilities. It allows the user to use the app offline without having to sign in. After signing in data that was kept locally is synced up. For a step-by-step guide, see [here](https://docs.powersync.com/integration-guides/supabase).
```

Chriztiaan commented:
There is currently a bug where if the user logs in immediately without creating data, existing data doesn't sync down. Looking into it.

Chriztiaan commented on demos/local-only-todolist/lib/powersync.dart:
```diff
@@ -0,0 +1,219 @@
+
+  if (!isSyncMode) {
+    await switchToSyncedSchema(db, getUserId());
+    // Without closing and reopening the database the list pages breaks if there is no data in the tables when logging in/signing up.
```
@stevensJourney @rkistner 
Before adding this `close/_openDb` snippet for switching to online from an **empty** local-only dataset, the list page's watch was broken.
I could see the data being synced down, but they weren't ending up in the watch stream even after the watch called got reinitialized. If I put a delay after the `await db.connect(connector: currentConnector);` on line 194 - the initial data would sync down, but even then any new data I created on the app or in supabase didn't trigger the watch with the latest data.

Is this workaround alright, or do we need to consider this a bug and not a limitation?

guillempuche commented:
@Chriztiaan, for defining locally a unique visitor id, it can be done with FingerprintJS https://dev.fingerprint.com/docs/identification-vs-fingerprintjs. I'm not sure if this can help you on the Typescript example.

Chriztiaan commented:
> @Chriztiaan, for defining locally a user id (fingerprint), it can be done with FingerprintJS https://dev.fingerprint.com/docs/identification-vs-fingerprintjs. I'm not sure if this can help you on the Typescript example.

This is cool! However I think we want to keep the demos as barebones/unopinionated as possible. I don't currently see a need to demonstrate anything more than a zeroed UUID.

Chriztiaan commented:
Using new version `sqlite_async` that has `refreshSchema`. Updated powersync package to allow invocation of `refreshSchema`. Added a [test](https://github.com/powersync-ja/powersync.dart/blob/12ddde23e97fdee3c2fcce95319e586cda4a0082/packages/powersync/test/offline_online_test.dart#L145), and using `refreshSchema` after the schema has been updated in the app.

benitav commented on demos/local-only-todolist/README.md:
```diff
@@ -0,0 +1,74 @@
+bucket_definitions:
+  user_lists:
+    # Separate bucket per todo list
+    parameters: select id as list_id from lists where owner_id = token_parameters.user_id
```
```suggestion
    parameters: select id as list_id from lists where owner_id = request.user_id()
```

Chriztiaan replied:
Thanks!

benitav commented on demos/local-only-todolist/README.md:
````diff
@@ -0,0 +1,74 @@
+```mermaid
+graph TD
+    K[Start] --> A[App is empty and local-only]
+    A --> B[User can create 0 or more todos that are stored local-only]
````
Somehow reading "create 0 todos" hurts my brain. Can we just say "User creates todos...". I feel if no data is created none of the syncing stuff is that relevant? Or is it?

Chriztiaan replied:
I can change it, I guess I wanted to make it clear that it doesn't matter whether the user creates 0 or more.

benitav commented on demos/local-only-todolist/lib/models/schema.dart:
```diff
@@ -0,0 +1,102 @@
+/// - `inactive_local_todos' and 'inactive_local_lists' refer to the local-only data, allowing
+///   for temporary storage or operations before syncing.
+///
+///  For an offline-to-online transition [switchToOnlineSchema] copies data so that it ends up in the upload queue.
```
Looks like `switchToOnlineSchema` was renamed to `switchToSyncedSchema`?

kobiebotha commented on demos/local-only-todolist/README.md:
```diff
@@ -0,0 +1,106 @@
+# PowerSync + Supabase Flutter Todo List App: Local-Only to Sync Mode Demo
+
+This demo app is an extension of the [Supabase Todo List App](../supabase-todolist/README.md) and demonstrates using the PowerSync Flutter client SDK to persist data locally without syncing, for users to use the app without having to register or sign in. It then demonstrates syncing this data to Supabase once the user registers at a later stage.
```
This demo app is an extension of the [Supabase Todo List App](../supabase-todolist/README.md) and demonstrates using the PowerSync Flutter client SDK to persist data locally without syncing. This lets users use the app without having to register or sign in.

kobiebotha commented on demos/local-only-todolist/README.md:
```diff
@@ -0,0 +1,106 @@
+
+This app uses [local-only](https://pub.dev/documentation/powersync/latest/powersync/Table/Table.localOnly.html) tables to persist data until the user has registered or signed in. Local-only tables do not log updates in an upload queue, avoiding any overhead or growth in database size.
+
+Once the user registers, the data is moved over to synced tables, at which point the data would be placed in the upload queue and will start syncing.
```
Once the user registers, the data is moved over to synced tables. The move operation also places the data in the upload queue.

kobiebotha commented on demos/local-only-todolist/README.md:
```diff
@@ -0,0 +1,106 @@
+
+# How this works
+
+This app uses [local-only](https://pub.dev/documentation/powersync/latest/powersync/Table/Table.localOnly.html) tables to persist data until the user has registered or signed in. Local-only tables do not log updates in an upload queue, avoiding any overhead or growth in database size.
```
I think we can call it "the" upload queue, not an upload queue

kobiebotha commented on demos/local-only-todolist/README.md:
```diff
@@ -0,0 +1,106 @@
+
+A barebones way to achieve the above is to store and query data from local-only tables before registration, copy this data to each corresponding synced table after registration and then store and query from the synced tables. This would look as follows:
+
+![diagram-0](./assets/local-only-readme-0.png)
```
I think it would be good to include the `copy data:` label here as well to indicate what is meant by the arrow

guillempuche commented:
I'm ready for Typescript!
This is a great feature for a consumer app. Fewer clicks for experiencing AHA moment, login is a big barrier
